db/db_impl/db_impl_debug.cc

```diff
@@ -321,5 +322,49 @@ size_t DBImpl::TEST_EstimateInMemoryStatsHistorySize() const {
   InstrumentedMutexLock l(&const_cast<DBImpl*>(this)->stats_history_mutex_);
   return EstimateInMemoryStatsHistorySize();
 }
+
+void DBImpl::TEST_VerifyNoObsoleteFilesCached(
+    bool db_mutex_already_held) const {
+  // This check is somewhat expensive and obscure to make a part of every
+  // unit test in every build variety. Thus, we only enable it for ASAN builds.
+  if (!kMustFreeHeapAllocations) {
+    return;
+  }
+
+  std::optional<InstrumentedMutexLock> l;
+  if (db_mutex_already_held) {
+    mutex_.AssertHeld();
+  } else {
+    l.emplace(&mutex_);
+  }
+
+  std::vector<uint64_t> live_files;
+  for (auto cfd : *versions_->GetColumnFamilySet()) {
+    if (cfd->IsDropped()) {
+      continue;
+    }
+    // Sneakily add both SST and blob files to the same list
+    cfd->current()->AddLiveFiles(&live_files, &live_files);
```

Comment:
Sorry, one more question here: unlike the previous logic, this only adds the live files from the current Version; is it possible that when this method is called, we still have e.g. iterators or compactions holding on to earlier Versions? (Obsolete files would be the ones that are not needed by _any_ Versions that are in use, not just the latest.)

Reply:
> This can only be called when purging of obsolete files has
> "settled," such as during parts of DB Close().

You can't close with open iterators, but I should be more specific.

Reply:
Added to T206372901

Reply:
Actually, failure to consider all the versions might explain the outstanding issue with SST files 👍